src/api/settings/getServerSettings.js

```diff
@@ -14,14 +14,28 @@ export type AuthenticationMethods = {
   ...
 };
 
-export type ExternalAuthenticationMethod = {|
-  name: string,
+type BaseExternalAuthenticationMethod = {|
   display_name: string,
   display_icon: string | null,
   login_url: string,
   signup_url: string,
 |};
 
+export type AppleExternalAuthenticationMethod = {|
+  name: 'apple',
+  apple_kid: string,
+  ...BaseExternalAuthenticationMethod,
+|};
+
+export type OtherExternalAuthenticationMethod = {|
+  name: string, // I'd like this to be "all strings except 'apple'"
```

Comment:
I think this is a reasonable wish and there unfortunately isn't a way to fulfill it in Flow's type system. (But I don't blame Flow, because I think it'd be hard to support this without making the type system a lot more complicated.) Here's a previous occasion where I've wanted the same thing:
```
commit f3c6ff71768774834cd3584b832ebb35419bd31c
Author: Greg Price <greg@zulip.com>
Date:   Thu Apr 4 15:21:50 2019 -0700

    persist: Write down types on our vendored redux-persist-migrate.
    
    This version involves a couple of compromises:
    
    * The types aren't generic, as the library itself actually is --
      rather they're specialized to our particular `Action` and
      `GlobalState` types.  (And, to play nice with our `Action` type,
      we get the `REHYDRATE` constant from our own `actionConstants`.)
    
      On my first attempt at this (back on 2019-01-08), I tried to type
      this properly generically, in a way that would make sense for the
      upstream library.  I didn't find a way to make it work -- in
      particular, to give an appropriate type for Action.  The type
      should be "either RehydrateAction, or else an object where `type`
      is *not* `REHYDRATE`"... and I think there there may not actually
      be a way in Flow's type system to write that generically,
      i.e. short of listing out every possible value for the `type`
      property.  Using our concrete `Action` type is basically a way of
      doing that.
```

In that instance, we could solve it by actually listing out every possible value for `type`, because ultimately it was all internal to the app. In this instance, that solution isn't available.

Reply:
Cool, thanks! I remember coming across that commit message recently and making the connection.

One nice thing about our current strategy of enumerating the Kandra-hosted domains in `src/config.js` is that we don't need to make a separate `AppleExternalAuthenticationMethod` type that has stuff like `apple_kid` or `team_id` or `app_id`. So `ExternalAuthenticationMethod` can just stay how it is.

So this problem just vanishes. 🎉